### PR TITLE
Make `SeasonalMovingAverageModel` and `DeadlineMovingAverageModel` to work with context

### LINKED ISSUE #905: Make `SeasonalMovingAverageModel` and children to work with context
### 🚀 Feature Request

We want to teach `SeasonalMovingAverageModel` to work with context to enable inference mode.

### Proposal

1. Make `SeasonalMovingAverageModel` the child of `NonPredictionIntervalContextRequiredAbstractModel` instead of `NonPredictionIntervalContextIgnorantAbstractModel`.
2. Change `NonPredictionIntervalContextIgnorantModelMixin` into `NonPredictionIntervalContextRequiredModelMixin`.
3. Fix `_SeasonalMovingAverageModel` to work with context.
4. Fix `PerSegmentModelMixin` to work with prediction size, there probably will be some problems.
5. Fix tests on `test_inference` to work with a new type of model.

### Test cases

1. Make sure current tests pass.
2. Modify tests for `SeasonalMovingAverageModel`.
3. Fix `test_inference` tests.

### Additional context

Waiting for #845 to close.

### LINKED ISSUE #907: Make `DeadlineMovingAverageModel` to work with context
### 🚀 Feature Request

We want to teach `DeadlineMovingAverageModel` to work with context to enable inference mode.

### Proposal

1. Make `DeadlineMovingAverageModel` the child of `NonPredictionIntervalContextRequiredAbstractModel` instead of `NonPredictionIntervalContextIgnorantAbstractModel`.
2. Change `NonPredictionIntervalContextIgnorantModelMixin` into `NonPredictionIntervalContextRequiredModelMixin`.
3. Fix `_DeadlineMovingAverageModel` to work with context.
4. Fix `PerSegmentModelMixin` to work with prediction size, there probably will be some problems.
5. Fix tests on `test_inference` to work with a new type of model.

### Test cases

1. Make sure current tests pass.
2. Modify tests for `DeadlineMovingAverageModel`.
3. Fix `test_inference` tests.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
1. Look at #905.
2. Look at #907.

## Closing issues
Closes #905.
Closes #907.

### CODE REVIEW

File: tests/test_models/test_simple_models.py [hunk 1]
```diff
@@ -39,29 +39,45 @@ def df():
...
-    future_ts = simple_df.make_future(future_steps=7)
-    res = sma_model.forecast(future_ts)
+    future_ts = simple_df.make_future(future_steps=7, tail_steps=sma_model.context_size)
+    res = sma_model.forecast(future_ts, prediction_size=7)
     res = res.to_pandas(flatten=True)
     assert not res.isnull().values.any()
     assert len(res) == 14
 
 
+def test_simple_model_forecaster_fail(simple_df):
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:49]: Strange behavior, you do the make_future with the context_size tail steps and you still get not enough context

File: tests/test_models/test_simple_models.py [hunk 2]
```diff
@@ -39,29 +39,45 @@ def df():
...
-    future_ts = simple_df.make_future(future_steps=7)
-    res = model.forecast(future_ts)
+    future_ts = simple_df.make_future(future_steps=7, tail_steps=model.context_size)
+    res = model.forecast(future_ts, prediction_size=7)
     res = res.to_pandas(flatten=True)
     assert not res.isnull().values.any()
     assert len(res) == 14
 
 
+def test_sdeadline_model_forecaster_fail(simple_df):
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:68]: May be also a test for `_get_context_begining`

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_simple_models.py (hunk 1 above)
Comment (Mr-Geekman) [reply tests/test_models/test_simple_models.py:49]: We set `window=1000` for `SeasonalMovingAverageModel `, we can't have enough context here.

File: etna/analysis/outliers/prediction_interval_outliers.py [hunk 3]
```diff
@@ -86,9 +86,10 @@ def get_anomalies_prediction_interval(
         deepcopy(ts_inner), prediction_interval=True, quantiles=[lower_p, upper_p]
     )
     for segment in ts_inner.segments:
-        segment_slice = prediction_interval[:, segment, :][segment]
-        anomalies_mask = (segment_slice["target"] > segment_slice[f"target_{upper_p:.4g}"]) | (
-            segment_slice["target"] < segment_slice[f"target_{lower_p:.4g}"]
+        predicted_segment_slice = prediction_interval[:, segment, :][segment]
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/analysis/outliers/prediction_interval_outliers.py:89]: Earlier we had in `segment_slice` original target column, but in this PR we made models to write their predictions over original values in in-sample case. So to get original values we have to access `ts_inner`.

File: tests/test_models/test_simple_models.py (hunk 2 above)
Comment (Mr-Geekman) [reply tests/test_models/test_simple_models.py:126]: Added.

File: etna/models/deadline_ma.py [hunk 4]
```diff
@@ -78,65 +78,80 @@ def fit(self, df: pd.DataFrame, regressors: List[str]) -> "_DeadlineMovingAverag
...
         """
-        timestamps = df["timestamp"]
-        index = pd.date_range(start=self.timestamps.iloc[0], end=timestamps.iloc[-1])
-        res = np.append(self.series.values, np.zeros(len(df)))
+        context_beginning = self._get_context_beginning(df=df, prediction_size=prediction_size)
+
+        df_history = df.iloc[:-prediction_size]
+        history_targets = df_history["target"]
+        history_timestamps = df_history["timestamp"]
+        future_timestamps = df["timestamp"].iloc[-prediction_size:]
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/models/deadline_ma.py]: Fixed.

File: etna/models/deadline_ma.py [hunk 5]
```diff
@@ -78,65 +78,80 @@ def fit(self, df: pd.DataFrame, regressors: List[str]) -> "_DeadlineMovingAverag
...
                 f"It uses only target series for predict/\n "
             )
-        targets = df["target"]
-        timestamps = df["timestamp"]
+
+        self._freq = freq
+
+        return self
+
+    def _get_context_beginning(self, df: pd.DataFrame, prediction_size: int):
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/models/deadline_ma.py]: Fixed.

File: tests/test_models/test_simple_models.py (hunk 1 above)
Comment (alex-hse-repository) [reply tests/test_models/test_simple_models.py:50]: I mean no mistake on the first glance, may be it is better to name the test "test_simple_model_forecaster_fail_not_enough_history"?

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**